### PR TITLE
WIP: initial_states returns all states at a time

### LINKED ISSUE #616: Initial states for recurrent bricks are not annotated with meaningful names.
See #605.

### PR DESCRIPTION
**Pull request by rizar**
Fixes #616, see discussion in #605.

This is one step forward towards a clearer definition of what the inputs and the outputs of an application method are. 

@dmitriy-serdyuk , you might be interested in this PR. Could you please review it?

### CODE REVIEW

File: blocks/bricks/attention.py [hunk 1]
```diff
@@ -389,12 +386,10 @@ def take_glimpses_inputs(self):
...
-    def initial_glimpses(self, name, batch_size, attended):
-        if name == "weighted_averages":
-            return tensor.zeros((batch_size, self.attended_dim))
-        elif name == "weights":
-            return tensor.zeros((batch_size, attended.shape[0]))
-        raise ValueError("Unknown glimpse name {}".format(name))
+    def initial_glimpses(self, batch_size, attended):
+        return {
+         "weighted_averages": tensor.zeros((batch_size, self.attended_dim)),
+         "weights": tensor.zeros((batch_size, attended.shape[0]))}
```
Comment (dmitriy-serdyuk) [inline blocks/bricks/attention.py]: I thought, it should be done using `outputs=[...]` option of the `application` decorator, returning a dictionary looks bulky and may be a source of errors due to typos.

File: blocks/bricks/recurrent.py [hunk 2]
```diff
@@ -30,18 +30,19 @@ class BaseRecurrent(Brick):
     has_bias = False
 
     @application
-    def initial_state(self, state_name, batch_size, *args, **kwargs):
-        r"""Return an initial state for an application call.
+    def initial_states(self, batch_size, *args, **kwargs):
+        r"""Return an initial states for an application call.
```
Comment (dmitriy-serdyuk) [inline blocks/bricks/recurrent.py]: > an

File: blocks/bricks/attention.py (hunk 1 above)
Comment (rizar) [inline blocks/bricks/attention.py]: Indeed! Let's say those were cosmic rays influencing  my brain.. I will definitely do it as you suggest.

File: blocks/bricks/recurrent.py [hunk 3]
```diff
@@ -179,11 +190,11 @@ def recurrent_apply(brick, application, application_call,
...
-                    # TODO init_func returns 2D-tensor, fails for iterate=False
-                    kwargs[state_name] = (
-                        brick.initial_state(state_name, batch_size,
-                                            *args, **kwargs))
-                    assert kwargs[state_name]
+                    try:
+                        kwargs[state_name] = initial_states[state_name]
+                    except KeyError:
+                        raise KeyError(
+                            "no initial state for '{}'".format(state_name))
```
Comment (dmitriy-serdyuk) [inline blocks/bricks/recurrent.py]: May be we should add a link to a recurrent brick here. Like `No initial state for {} in brick {}`.

**Comment by dmitriy-serdyuk** [comment]
Nice work!

The only comment is that you should output tuples everywhere instead of dictionaries.

File: blocks/bricks/recurrent.py [hunk 4]
```diff
@@ -179,11 +190,11 @@ def recurrent_apply(brick, application, application_call,
                             kwargs[state_name](state_name, batch_size,
                                                *args, **kwargs))
                 else:
-                    # TODO init_func returns 2D-tensor, fails for iterate=False
```
Comment (dmitriy-serdyuk) [inline blocks/bricks/recurrent.py]: Can you open an issue for it?

Comment (dmitriy-serdyuk) [inline blocks/bricks/recurrent.py]: Oops, I meant another one TODO.

File: blocks/search.py [hunk 5]
```diff
@@ -95,11 +95,10 @@ def _compile_context_computer(self):
             self.inputs, self.contexts, on_unused_input='ignore')
 
     def _compile_initial_state_computer(self):
-        initial_states = [
-            self.generator.initial_state(
-                name, self.beam_size,
+        # TODO: should be now extractable from the computation graph
```
Comment (dmitriy-serdyuk) [inline blocks/search.py]: I created #719 for it.

**Comment by rizar** [comment]
Seems like I addressed all your comments. Would you mind taking one more look?

**Comment by dmitriy-serdyuk** [comment]
LGTM

Is there any difference returning a tuple or a list?

**Comment by rizar** [comment]
In the contexts of Blocks application methods not so much, I guess, since we anyway make it a list, see blocks/bricks/base.py:303

**Pull request merged**